openmdao/jacobians/assembled_jacobian.py

```diff
@@ -327,14 +333,90 @@ def _apply(self, d_inputs, d_outputs, d_residuals, mode):
+
+                if ext_mtx is not None and \
+                   len(d_inputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    # Masking
+                    cache_key = tuple(d_inputs._names)
+                    if cache_key not in self._mask_caches:
+                        self._create_mask_cache(d_inputs, cache_key, ext_mtx)
+
+                    mask = self._mask_caches.get(cache_key)
+                    if mask is not None:
+                        inputs_masked = np.ma.array(d_inputs.get_data(), mask=mask)
+
+                        # Use the special dot product function from masking module so that we
+                        # ignore masked parts.
+                        d_residuals.iadd_data(np.ma.dot(ext_mtx._matrix, inputs_masked))
+
+                    else:
+                        d_residuals.iadd_data(ext_mtx._prod(d_inputs.get_data(), mode, None))
+
             else:  # rev
                 dresids = d_residuals.get_data()
-                d_outputs.iadd_data(int_mtx._prod(dresids, mode, int_ranges))
-                if ext_mtx is not None:
-                    d_inputs.iadd_data(ext_mtx._prod(dresids, mode, None))
+                if len(d_outputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    d_outputs.iadd_data(int_mtx._prod(dresids, mode, int_ranges))
+
+                if ext_mtx is not None and \
+                   len(d_inputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    # Masking
+                    cache_key = tuple(d_inputs._names)
+                    if cache_key not in self._mask_caches:
+                        self._create_mask_cache(d_inputs, cache_key, ext_mtx)
+
+                    mask_cols = self._mask_caches.get(cache_key)
+                    if mask_cols is not None:
+
+                        # Mask need to be applied to ext_mtx so that we can ignore multiplication
+                        # by certain columns.
+                        mask = np.zeros(ext_mtx._matrix.T.shape, dtype=np.bool)
+                        mask[mask_cols, :] = True
+                        masked_mtx = np.ma.array(ext_mtx._matrix, mask=mask, fill_value=0.0)
+
+                        masked_product = np.ma.multiply(masked_mtx.T, dresids).flatten()
+
+                        for set_name, data in iteritems(d_inputs._data):
+                            data += np.ma.filled(masked_product)
+
+                    else:
+                        d_inputs.iadd_data(ext_mtx._prod(dresids, mode, None))
+
+    def _create_mask_cache(self, d_inputs, cache_key, ext_mtx):
+        """
+        Create masking array for d_inputs vector.
+
+        Parameters
+        ----------
+        d_inputs : Vector
+            The inputs linear vector.
+        cache_key : tuple
+            Hashable unique key, from d_inputs._names
+        ext_mtx : Matrix
+            External matrix
+        """
+        masked = [name for name in d_inputs._views if name not in cache_key]
```

Comment:
This loop should be flipped since checking containment in d_inputs_views (a dict) will be much faster than checking containment in cache_key (a tuple).

Reply:
Fixed